src/vcpkg/sourceparagraph.cpp

```diff
@@ -708,142 +787,303 @@ namespace vcpkg
     };
     ContactsDeserializer ContactsDeserializer::instance;
 
-    static constexpr StringLiteral EXPRESSION_WORDS[] = {
-        "WITH",
-        "AND",
-        "OR",
-    };
-    static constexpr StringLiteral VALID_LICENSES[] =
+    static constexpr StringLiteral VALID_LICENSES[] = {
 #include "spdx-licenses.inc"
-        ;
-    static constexpr StringLiteral VALID_EXCEPTIONS[] =
-#include "spdx-licenses.inc"
-        ;
+    };
+    static constexpr StringLiteral VALID_EXCEPTIONS[] = {
+#include "spdx-exceptions.inc"
+    };
 
-    // We "parse" this so that we can add actual license parsing at some point in the future
-    // without breaking anyone
-    struct LicenseExpressionDeserializer : Json::IDeserializer<std::string>
+    // The "license" field; either:
+    // * a string, which must be an SPDX license expression.
+    //   EBNF located at: https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/manifest-files.md#license
+    // * `null`, for when the license of the package cannot be described by an SPDX expression
+    static const StringLiteral* case_insensitive_find(View<StringLiteral> lst, StringView id)
     {
-        virtual StringView type_name() const override { return "an SPDX license expression"; }
+        return Util::find_if(lst, [id](StringLiteral el) { return Strings::case_insensitive_ascii_equals(id, el); });
+    }
+    static constexpr bool is_idstring_element(char ch) { return Parse::is_alphanumdash(ch) || ch == '.'; }
 
-        enum class Mode
+    std::string parse_spdx_license_expression(StringView sv, std::vector<ParseSpdxLicenseError>& errors)
```

Comment:
Why not use our existing parser framework for this (`Parse::Parserbase`)?

I see two differences compared to the code you have here:
1. Localization - we must eventually add Localization to ParserBase for all the other parsers in the product, so this doesn't seem like a good reason to do things completely ad-hoc here. Either this new code should use the existing framework (and get localized when we localize every other parser) or we should localize ParserBase first in a separate PR and then add this code.
2. Warnings - adding warnings to the existing parsing framework seems obviously good and not long or complex.

Reply:
I didn't think about it until I was very close to done with this implementation. I can switch it over, I think, pretty easily tho.

Reply:
hmm, now that I'm looking at this, it seems like it'd be quite difficult to push this parser into the old ParserBase model; I would prefer to rewrite the model from scratch, and then switch stuff over as it comes up. I also don't want to add new un-localized code.